File: src/leetcode/2026/03/Solution696.cpp
#include <string>
using namespace std;

class Solution {
public:
    int countBinarySubstrings(string s) {
        int prev = 0, cur = 1, ans = 0;

        for (int i=1; i<s.size(); i++) {
            if (s[i] == s[i-1]) {
                cur++;
            } else {
                prev = cur;
                cur = 1;
            }
            if (prev >= cur) {
                ans++;
            }
        }

        return ans;
    }
};

int main() {
    Solution sol;

    string s = "00110011";
    int ans = sol.countBinarySubstrings(s);
    printf("%d\n", ans);

    s = "10101";
    ans = sol.countBinarySubstrings(s);
    printf("%d\n", ans);
}

File: src/leetcode/2026/03/Solution696.java
public class Solution696 {
    public int countBinarySubstrings(String s) {
        int prevRunLength = 0;
        int curRunLength = 1;
        int count = 0;

        for (int i = 1; i < s.length(); i++) {
            if (s.charAt(i) == s.charAt(i - 1)) {
                curRunLength++;
            } else {
                prevRunLength = curRunLength;
                curRunLength = 1;
            }
            if (prevRunLength >= curRunLength) {
                count++;
            }
        }

        return count;
    }

    public static void main(String[] args) {
        Solution696 solution = new Solution696();

        String s = "00110011";
        int result = solution.countBinarySubstrings(s);
        System.out.println(result); // Output: 6

        s = "10101";
        result = solution.countBinarySubstrings(s);
        System.out.println(result); // Output: 4
    }
}
